core/src/main/scala/kafka/server/DynamicBrokerConfig.scala

```diff
@@ -0,0 +1,280 @@
+    updateDefaultConfig(adminZkClient.fetchEntityConfig(ConfigType.Broker, ConfigEntityName.Default))
+    updateBrokerConfig(brokerId, adminZkClient.fetchEntityConfig(ConfigType.Broker, brokerId.toString))
+    updateCurrentConfig()
+  }
+
+  def config: KafkaConfig = CoreUtils.inReadLock(lock) {
+    currentConfig
+  }
+
+  def addReconfigurable(reconfigurable: Reconfigurable): Unit = CoreUtils.inWriteLock(lock) {
+    require(reconfigurable.reconfigurableConfigs.asScala.forall(allDynamicConfigs.contains))
+    reconfigurables += reconfigurable
+  }
+
+  def removeReconfigurable(reconfigurable: Reconfigurable): Unit = CoreUtils.inWriteLock(lock) {
+    reconfigurables -= reconfigurable
+  }
+
+  private[server] def updateBrokerConfig(brokerId: Int, persistentProps: Properties): Unit = CoreUtils.inWriteLock(lock) {
+    val props = fromPersistentProps(persistentProps)
+    validateConfigTypes(Some(brokerId), props, logError = true)
+    dynamicBrokerConfigs.clear()
+    dynamicBrokerConfigs ++= props.asScala
+    updateCurrentConfig()
+  }
+
+  private[server] def updateDefaultConfig(persistentProps: Properties): Unit = CoreUtils.inWriteLock(lock) {
+    val props = fromPersistentProps(persistentProps)
+    validateConfigTypes(None, props, logError = true)
+    dynamicDefaultConfigs.clear()
+    dynamicDefaultConfigs ++= props.asScala
+    updateCurrentConfig()
+  }
+
+  private[server] def validateAndConvertForPersistence(brokerId: Option[Int], props: Properties): Properties = CoreUtils.inReadLock(lock) {
+    validateConfigTypes(brokerId, props, logError = false)
+    val newProps = mutable.Map[String, String]()
+    newProps ++= staticBrokerConfigs
+    brokerId match {
+      case Some(_) =>
+        overrideProps(newProps, dynamicDefaultConfigs)
+        overrideProps(newProps, props.asScala)
+      case None =>
+        overrideProps(newProps, props.asScala)
+        overrideProps(newProps, dynamicBrokerConfigs)
+    }
+    processConfig(newProps, validateOnly = true)
+    toPersistentProps(props)
+  }
+
+  private def validateConfigTypes(brokerId: Option[Int], props: Properties, logError: Boolean): Unit = {
+    try {
+      val securityConfigsWithoutPrefix = dynamicSecurityConfigs.filter(props.containsKey)
+      if (securityConfigsWithoutPrefix.nonEmpty)
+          throw new ConfigException(s"Invalid configs $securityConfigsWithoutPrefix: security configs can be dynamically updated only using listener prefix")
+      val baseProps = new Properties
+      props.asScala.foreach {
+        case (listenerConfigRegex(baseName), v) => baseProps.put(baseName, v)
+        case (k, v) => baseProps.put(k, v)
+      }
+      DynamicConfig.Broker.validate(baseProps)
+    } catch {
+      case e: Exception =>
+        if (logError)
+          error(s"Dynamic default broker config is invalid: $props, these configs will be ignored", e)
+        throw e
+    }
+  }
+
+  private def configToMap(config: Map[String, _]): immutable.Map[String, String] = {
```

Comment:
I don't think we would want to use this anywhere else since it converts passwords to the actual value - we want to avoid doing this elsewhere. Have added a comment to clarify that.